<a href="https://colab.research.google.com/github/saipravalika1224/Data-Cleaning-task/blob/main/Used_Car_Price_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:

       import pandas as pd
import numpy as np

# Load your dataset (Replace 'cars.csv' with your file path or URL)
# df = pd.read_csv('your_dataset.csv')

# Load your dataset
df = pd.read_csv('/used_cars_messy.csv')

# Exploration
print("--- Dataset Info ---")
df.info()
print("\n--- Descriptive Statistics ---")
display(df.describe())
print(f"\nShape: {df.shape}")

# Identify 3 Issues (Observations to report):
# 1. Null values in 'selling_price' or feature columns.
# 2. 'mileage' is likely an object/string (e.g., '15 kmpl') instead of float.
# 3. 'brand' names might have inconsistent casing or leading spaces.

--- Dataset Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15661 entries, 0 to 15660
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         15661 non-null  int64  
 1   car_name           15661 non-null  object 
 2   brand              15661 non-null  object 
 3   model              15661 non-null  object 
 4   vehicle_age        15661 non-null  int64  
 5   km_driven          15661 non-null  object 
 6   seller_type        15661 non-null  object 
 7   fuel_type          15661 non-null  object 
 8   transmission_type  15661 non-null  object 
 9   mileage            14376 non-null  object 
 10  engine             13792 non-null  float64
 11  max_power          14862 non-null  float64
 12  seats              15218 non-null  float64
 13  selling_price      15503 non-null  float64
dtypes: float64(4), int64(2), object(8)
memory usage: 1.7+ MB

--- Descriptive Statistics ---


,Unnamed: 0,vehicle_age,engine,max_power,seats,selling_price
count,15661.000000,15661.000000,13792.000000,14862.000000,15218.000000,1.550300e+04
mean,9808.189643,6.036843,1480.489342,100.513582,5.324418,1.032764e+06
std,5642.981540,3.015072,511.446057,42.854632,0.805993,1.607380e+07
min,0.000000,0.000000,793.000000,38.400000,0.000000,5.000000e+02
25%,4901.000000,4.000000,1197.000000,74.000000,5.000000,3.850000e+05
50%,9858.000000,6.000000,1248.000000,88.500000,5.000000,5.550000e+05
75%,14666.000000,8.000000,1582.000000,117.300000,5.000000,8.250000e+05
max,19543.000000,29.000000,6592.000000,626.000000,9.000000,1.000000e+09



Shape: (15661, 14)


Task 1: Data Quality Issues Identified
After exploring the used_cars_messy.csv dataset, I identified the following three issues:
1.	Missing Values: The mileage and selling_price columns have significant null values (only 14,376 non-null rows out of 15,661).
2.	Incorrect Data Types: The mileage and km_driven columns are stored as objects (text) because they contain units like "kmpl" and commas. They must be numeric for analysis.
3.	Logical Errors: The seats column contains a minimum value of 0.0, which is an impossible value for a car and indicates a data entry error.

In [ ]:
# 1. Drop rows where selling_price is missing
df.dropna(subset=['selling_price'], inplace=True)

# 2. Extract numbers from 'mileage' and 'km_driven'
df['mileage'] = df['mileage'].astype(str).str.extract('(\d+\.\d+|\d+)').astype(float)
df['km_driven'] = df['km_driven'].astype(str).str.replace(',', '').str.extract('(\d+)').astype(float)

# 3. Impute (Fill) missing values
df['mileage'] = df['mileage'].fillna(df['mileage'].median())
df['engine'] = df['engine'].fillna(df['engine'].median())

# 4. Clean 'brand' column
df['brand'] = df['brand'].str.strip().str.lower()

# 5. Remove duplicates
df.drop_duplicates(inplace=True)

print("Cleaning complete! New Shape:", df.shape)

Cleaning complete! New Shape: (15322, 14)


<>:5: SyntaxWarning: invalid escape sequence '\d'
<>:6: SyntaxWarning: invalid escape sequence '\d'
<>:5: SyntaxWarning: invalid escape sequence '\d'
<>:6: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipykernel_6636/3815597684.py:5: SyntaxWarning: invalid escape sequence '\d'
  df['mileage'] = df['mileage'].astype(str).str.extract('(\d+\.\d+|\d+)').astype(float)
/tmp/ipykernel_6636/3815597684.py:6: SyntaxWarning: invalid escape sequence '\d'
  df['km_driven'] = df['km_driven'].astype(str).str.replace(',', '').str.extract('(\d+)').astype(float)


Task 2: Data Cleaning Summary
I performed the following cleaning steps:
* Dropped rows where the target variable (selling_price) was missing.
* Extracted numeric values from text columns using Regex and converted them to floats.
* Handled missing values in mileage and engine by imputing with the median.
* Standardized the brand column and removed duplicate rows.
* Final Dataset Shape: (15322, 14)

In [ ]:
from sklearn.metrics import mean_absolute_error

# Predict the average (mean) price for every car
y_true = df['selling_price']
prediction = y_true.mean()
y_pred = [prediction] * len(y_true)

# Calculate MAE
baseline_mae = mean_absolute_error(y_true, y_pred)

print(f"Mean Selling Price: {prediction:.2f}")
print(f"Baseline MAE: {baseline_mae:.2f}")

Mean Selling Price: 1036662.05
Baseline MAE: 867477.39


Task 3: Baseline Metric
* Mean Selling Price: 1,036,662.05
* Baseline Mean Absolute Error (MAE): 867,477.39
* This MAE represents the average error if we simply predicted the mean price for every car.